Implement useful checks from kvalobs

intarga opened the pull request:
Closes #34 and #1

- [x] range check
- [x] range check direction
    - prototype written (range_check_wind_direction) but need to verify Y param and -3.0 special case with Line
- [x] logger_t
    - implemented as special_values_check
- [x] freeze check
    - implemented as flatline_check
- [x] geonor_drift
    - implemented as monotonic_increase_check
    - Why does it have 4 different flags in the log, when the linked code only returns 1 possible flag, and has no parameters
    - Why the hardcoded values 100 and 0.7?
    - Why does this strictly apply to 25 values in a row? 25 seems like a strange number
    - It only returns 1 flag. Does that apply to all 25 values?
    - Verify I got the pass and fail flags the right way around
    - Unsure how to apply this to a whole DataCache. Rolling windows of 25 values, or chunks?
- [x] humidity_range
    - implemented as range_check_humidity, but need to verify flag on correction
    - what to do about the warnings for 5-15%? Should we put a warn case in this test, or run a separate range check to test for that? 
~~- [ ] freeze_check_uu~~
    - The code for this really didn't make sense at all, and has a very obvious mistake in it. I suspect this is based on an early iteration of freeze_check, and then they updated freeze_check while leaving this unchanged. It seems like the only actually relevant part of this is that they exit early if humidity is > 95. I suggest that instead of having a special check for that we just use the regular flatline_check + a range check and handle it in the combi
- [x] gt
    - implemented at greater_than
    - Verify I got the pass and fail flags the right way around
- [x] max_gt
    - implemented as max_greater_than_single, the other 3 tests below follow the same naming convention
    - flagging conditions for missing data changed significantly, my version does not flag as pass if there's any missing data (but it can still flag as fail). kvalobs flags only one data point in the sequence, and is inconsistent about which one, I just flag everything together, open to changing that if a clear and coherent argument can be given.
- [x] min_gt
- [x] min_lt
- [x] max_lt
- [x] prognostic_space_check (compare_with_model)
    - implemented as not_equal
- [x] remove series/spatial cache at the olympian level
- [x] unit tests

~~- [x] badweather_highvisibility (implemented as range_check_pair)~~
~~- [ ] mediumclouds_low
    - prototype implemented as cloud_consistency_check, but need help to verify types, and provide a better description
    - this check only seems to look at clouds, but in the doc it's marked as flagging pressure?~~
~~- [ ] cloudbase_hi
    - prototype implemented as cloud_consistency_check2, but same comments apply as mediumclouds_low.
    - Additionally, I need help picking better names here
    - as above, I am concerned that this is used to QC pressure but doesn't actually look at pressure~~
~~- [x] lt_eq
    - implemented as lower_limit_special_value_pair
    - as above, I am concerned that this is used to QC pressure but doesn't actually look at pressure~~
~~- [x] gt_eq
    - implemented as upper_limit_special_value_pair
    - as above, I am concerned that this is used to QC pressure but doesn't actually look at pressure~~

intarga committed:
implement range_check

intarga committed:
implement freeze check

intarga committed:
implement special_values_check

intarga committed:
implement range_check_wind_direction

intarga committed:
implement prototype aggregate_less_than_set

intarga committed:
implement range_check_pair

intarga committed:
implement range_check_humidity

intarga committed:
implement cloud_consistency_check

intarga committed:
implement cloud_consistency_check2

intarga committed:
implement lower_limit_special_value_pair

intarga committed:
implement upper_limit_special_value_pair

linebaserud commented:
My initial thoughts on these questions:

range check direction
prototype written (range_check_wind_direction) but need to verify Y param and -3.0 special case with Line

I do not understand this either. Y wind speed? What is special with -3.0?? Maybe PiM knows who to ask? 


 logger_t
what is this??? the code looks like it checks control info and a bunch of special values, but it has a comment that says it's a range check?

This should just be our special_values_check, right? Maybe it says so since it is using the same flag, fr/controlinfo(1). I find explanations like "Gjoer tester paa 9999-verdier og returner en ny verdi for grenseverdi-flagg" and "fr=7 Kontrollert. Funnet å svare til spesialverdi som betyr manglende")
From the code it looks like Øystein Lie is the person to ask for more info.

 min_gt
prototype written (aggregate_less_than_set) but the test seems flawed to me. I don't think it should be able to pass the aggregate if there are missing values in the set (should be at best DataMissing), and if the condition fails, I think it should fail all elems of the set that fail the condition, not just the lowest. Also what is the adjustment? Where does that magic number come from?

I do not understand this after the first quick look. 

 geonor_drift
No explanation at all for this test. it checks if a set of 25 values are monotonically increasing, and if the difference between the first and last is in a hardcoded range. Struggling to see the rationale behind this one. Additionally why does it have 4 different flags in the log, when the linked code only returns 1 possible flag, and has no parameters

Geonor should be the precip buckets that fills with water as it rains. They are continuously weighed to measure the precipitation. -> values should be monotonically increasing. During the summer I guess some water can evaporate away. I do not know way they use an extra range check here. I can only see 1 flag in the log, not 4.

On the consistency checks (badweather_highvisibility, mediumclouds_low, cloudbase_hi, lt_eq, gt_eq, maxtemp_nosign, range_max_min):
I agree that these checks do not seem to look at pressure!
These seem to be consistency checks that have names that uses numbers equal to the paramids used by pressure variables:

consistency checks clouds uses 170, 171, 172, 173, and 175:
50 	QC1-2-170 - nolowclouds_lowcloudheight2
445 	QC1-2-171 - Cu_low
253 	QC1-2-172 - mediumclouds_low
55 	QC1-2-173 - cloudbase_hi
113 	QC1-2-175 - lt_eq

and this one related to seatemp uses 179:
822 	QC1-2-179 - flowrate_for_seatemp

and something related to temp uses 177 and 178 in the naming:
1 	QC1-2-177b - maxtemp_nosign
1 	QC1-2-177 - maxtemp_nosign
18 	QC1-2-178 - range_max_min

also visibility/weather uses 176:
14737 	QC1-6-176 - gt_eq

Here I find the consistency checks for pressure variables:
https://gitlab.met.no/groups/met/obsklim/bakkeobservasjoner/data-og-kvalitet/kvalobs/-/wikis/kvalobs/qc12#air-pressure
And this is the related statistics from the kvalobs log files (the gt_eq test are in fact used, but now with more relevant input):
58096 		CHECK: QC1-2-21 - pressure_nottendency (AA = null & PP ≥ 0)
254 		CHECK: QC1-2-22 - tendency_notpressure (AA = 4 & PP ≠ 0)
23262 		CHECK: QC1-2-23 - gt_eq (AA = 4 & PP ≠ 0)
537 		CHECK: QC1-2-24 - eq_between (PP = 0 & 1 ≤ AA ≤ 3)
1619 		CHECK: QC1-2-25 - eq_between (PP = 0 & 6 ≤ AA ≤ 8)
0 		CHECK: QC1-2-26 - pressurereduction 
(Check on station pressure reduction. The check only applies to stations where reduction is calculated manually and 950 hPa ≤ PR ≤ 1050 hPa. The inconsistency is: | PR-P1 | > ε, where ε is instrument specific)
(125783 FLAG=1, 143487 EMPTY RESULTS)
73298 		CHECK: QC1-2-27.b - tendency (abs( PP - |PO - PO(-3h)| ) > 0.3 hPa)

intarga commented:
@linebaserud 

> I do not understand this either. Y wind speed? What is special with -3.0?? Maybe PiM knows who to ask?

I think Y is wind speed, since it makes sense that wind direction would become meaningless when wind speed is 0, but it's not mentioned in the description. As for the -3.0 it seems like special handling for some specific sensor? 

> I do not understand this after the first quick look.

I think this is maybe easier to explain in person and with a diagram, we can bring it up at the next meeting

> I can only see 1 flag in the log, not 4.

I meant that in your document it is listed as 4 entries:
```
263 	1719 		QC1-3a-104drift.501 - geonor_drift
295 	1012 		QC1-3a-104drift.502 - geonor_drift
373 	176 		QC1-3a-104drift.330 - geonor_drift
...
466 	24 		QC1-3a-104drift.342 - geonor_drift
```
But it's unclear to me what (if any) difference there could be between these checks, as the linked code has only 1 function which does not take in any params

> I agree that these checks do not seem to look at pressure!
These seem to be consistency checks that have names that uses numbers equal to the paramids used by pressure variables:

Ok so those tests leaked into the list because the regex you used on the logs wasn't quite right? Or is it a mistake that they're named like that? In that case it seems like I shouldn't implement them, right? I feel kind of sad that I spent so long puzzling about this when it never made sense.

You linked the new set of consistency checks that actually are used for pressure, do you want me to implement these? are they actually flagging anything?

linebaserud commented:
@intarga 

>>  I do not understand this either. Y wind speed? What is special with -3.0?? Maybe PiM knows who to ask?

> I think Y is wind speed, since it makes sense that wind direction would become meaningless when wind speed is 0, but it's not mentioned in the description. As for the -3.0 it seems like special handling for some specific sensor?

Sounds reasonable. I will put it on my list to try to ask some people to verify.

> > I do not understand this after the first quick look.

> I think this is maybe easier to explain in person and with a diagram, we can bring it up at the next meeting

Perfect. It is now in the agenda for the next meeting.

>> I can only see 1 flag in the log, not 4.

> I meant that in your document it is listed as 4 entries:
> 263 	1719 		QC1-3a-104drift.501 - geonor_drift
> 295 	1012 		QC1-3a-104drift.502 - geonor_drift
> 373 	176 		QC1-3a-104drift.330 - geonor_drift
> ...
> 466 	24 		QC1-3a-104drift.342 - geonor_drift
> But it's unclear to me what (if any) difference there could be between these checks, as the linked code has only 1 function which does not take in any params

It should be the same test (https://gitlab.met.no/met/obsklim/bakkeobservasjoner/data-og-kvalitet/kvalobs/kvoss_intern/-/blob/master/kvmeta/algorithms/geonor_drift.pl?ref_type=heads), but run for the different typeids (501,502,330,...,342). 

>> I agree that these checks do not seem to look at pressure!
>> These seem to be consistency checks that have names that uses numbers equal to the paramids used by pressure variables:

> Ok so those tests leaked into the list because the regex you used on the logs wasn't quite right? Or is it a mistake that they're named like that? In that case it seems like I shouldn't implement them, right? I feel kind of sad that I spent so long puzzling about this when it never made sense.

> You linked the new set of consistency checks that actually are used for pressure, do you want me to implement these? are they actually flagging anything?

Correct, leaked in w/ the regex. Maybe that is showing us that naming most of the tests with the paramid, but also naming lists of tests with the same numbers can be confusing... 
We should not implement them for pressure, however, these consistency tests do flag observations so we will need them for checking the cloud observations etc. The document I made for you shows the kvalobs checks for the main parameters (temperature, precip, wind, rh, pressure) that we wanted to start with, but in the end we need to check also all the other observations that kvalobs checks. 
The new set is flagging, except for CHECK: QC1-2-26 - pressurereduction. The number of flagged observations is the number just before each test that I listed in my previous comment.

intarga committed:
replace SeriesCache and SpatialCache with DataCache

intarga committed:
fix spike_check to use DataCache

intarga committed:
fix flatline_check to use DataCache

intarga committed:
fix step_check to use DataCache

intarga committed:
fix special_values_check to use DataCache

intarga committed:
fix range_check to use DataCache

intarga committed:
remove checks that shouldn't have been in Line's list

intarga committed:
fix range_check_humidity to use DataCache

intarga committed:
fix range_check_wind_direction to use DataCache

intarga committed:
fix buddy_check to use DataCache

intarga committed:
fix sct to use DataCache

intarga committed:
add missing clones on strings in return vecs

intarga committed:
add missing docs and debug impls

intarga committed:
fix clippy lints

intarga committed:
fix misplaced arguments in sct and buddy_check tests

intarga committed:
fix doctest

intarga committed:
reorganise module hierarchy

intarga committed:
implement monotonic_increase_check

intarga committed:
implement greater_than.rs

intarga committed:
better naming and logic in aggregate_less_than_set (now single_greater_than_element)

intarga committed:
change the naming and logic of the consistency check again

I scrutinised the kvalobs code, and it's not very consistent, but this
best matches the behaviour in a consistent way.

The only thing left to consider is whether we should return separate
flags for the sequence values. kvalobs flags only one of them (and is
rather inconsistent about which one some flag the min/max while others
flag only the first value weirdly). I will consult with the others in
Confident WP5 and revisit this based on feedback.

intarga committed:
implement max_greater_than_single

intarga committed:
implement min_less_than_single

intarga committed:
implement max_less_than_single

intarga committed:
unit test single checks

intarga committed:
unit test series checks

intarga committed:
unit test consistency checks

intarga committed:
implement not_equal

Lun4m commented on src/checks/consistency/greater_than.rs:
```diff
@@ -0,0 +1,32 @@
+/// Returns [`Flag::DataMissing`] if either datum is missing,
+/// [`Flag::Fail`] if datum1 + datum1_correction > datum2,
+/// [`Flag::Pass`] otherwise.
+pub fn greater_than(datum1: Option<f32>, datum2: Option<f32>, datum1_correction: f32) -> Flag {
```
We may need to make these functions generic in the future, so maybe it's worth it adding a TODO?

intarga replied:
I think it's not that helpful because needing to do this will be prompted externally. It's not that relevant just when looking at the code

Lun4m commented on src/checks/consistency/max_greater_than_single.rs:
```diff
@@ -0,0 +1,82 @@
+        });
+    let max = match max {
+        Some(value) => value,
+        // if min is None at this point, then all the elements of the sequence were missing...
```
```suggestion
        // if max is None at this point, then all the elements of the sequence were missing...
```

Lun4m commented on src/checks/consistency/max_greater_than_single.rs:
```diff
@@ -0,0 +1,82 @@
+/// If this invariant is broken (i.e the maximum of the sequence, plus the adjustment, is greater
+/// than the single value), we return [`Flag::Fail`].
+/// Else, if any of the elements is missing, we return [`Flag::DataMissing`], as we cannot be sure
+/// a missing data point did not violate the invariant.
```
Is this how it is implemented in Kvalobs? I don't know, I feel like maybe it should return a different flag?

Lun4m commented on src/checks/consistency/max_greater_than_single.rs:
```diff
@@ -0,0 +1,82 @@
+    #[test]
+    fn test_max_greater_than_single() {
+        assert_eq!(
+            max_greater_than_single(Some(1.), &vec![Some(1.), Some(0.), Some(0.)], 0.2),
+            Flag::Fail
+        );
+        assert_eq!(
+            max_greater_than_single(Some(1.), &vec![Some(1.), Some(0.), Some(0.)], -0.2),
+            Flag::Pass
+        );
+        assert_eq!(
+            max_greater_than_single(Some(1.), &vec![Some(1.), None, Some(0.)], -0.2),
+            Flag::DataMissing
+        );
+        assert_eq!(
+            max_greater_than_single(Some(1.), &vec![Some(1.), None, Some(0.)], 0.2),
```
Minor, but clippy says we can use directly the slices (i.e. we can remove `vec!`)

intarga replied:
Oh, weird that CI didn't catch this? I'll look into why

intarga replied:
Apparently we need `cargo clippy --tests` to get it to lint inside `[#cfg(test)]` blocks, will fix that too

intarga replied:
fixed

Lun4m commented on src/checks/consistency/max_less_than_single.rs:
```diff
@@ -0,0 +1,82 @@
+        });
+    let max = match max {
+        Some(value) => value,
+        // if min is None at this point, then all the elements of the sequence were missing...
```
```suggestion
        // if max is None at this point, then all the elements of the sequence were missing...
```

Lun4m commented on src/checks/consistency/max_less_than_single.rs:
```diff
@@ -0,0 +1,82 @@
+        assert_eq!(
+            max_less_than_single(Some(1.), &vec![Some(1.), None, Some(0.)], 0.2),
+            Flag::DataMissing
+        );
```
Same as in `max_greater_than_single`

Lun4m commented on src/checks/consistency/min_greater_than_single.rs:
```diff
@@ -0,0 +1,82 @@
+        assert_eq!(
+            min_greater_than_single(Some(1.), &vec![Some(1.), None, Some(2.)], 0.2),
+            Flag::Fail
+        );
```
Same as in `max_greater_than_single`

Lun4m commented on src/checks/consistency/min_less_than_single.rs:
```diff
@@ -0,0 +1,82 @@
+        assert_eq!(
+            min_less_than_single(Some(1.), &vec![Some(1.), None, Some(2.)], 0.2),
+            Flag::DataMissing
+        );
```
Same as in `max_greater_than_single`

Lun4m commented on src/checks/series/flatline_check.rs:
```diff
@@ -0,0 +1,107 @@
+use crate::{DataCache, Error, Flag};
+
+/// Timeseries QC test that checks for streaks of repeating values.
```
```suggestion
/// Timeseries QC check that looks for streaks of repeating values.
```
Since we said we should use `check` instead of `test`, maybe you also want to leave out `QC`?

Lun4m commented on src/checks/series/flatline_check.rs:
```diff
@@ -0,0 +1,107 @@
+    if data.contains(&None) {
+        return Flag::DataMissing;
+    }
+    let data: Vec<f32> = data.iter().map(|opt| opt.unwrap()).collect();
```
I don't think we need to create a new vector here?
```suggestion
```

intarga replied:
fixed

Lun4m commented on src/checks/series/flatline_check.rs:
```diff
@@ -0,0 +1,107 @@
+/// else Flag::Pass.
+///
+/// As (`num_points` - 1) predecessors to each observation are needed, the [`SeriesCache`] provided must have
+/// `num_leading_points` >= `num_points` - 1.
```
Minor rewording and I added a section explaining `DataCache`, but maybe it should be moved close to the struct definition? It might be also important to point out that the length of the returned inner vectors is equal to (`cache.data.len() - cache.num_leading_points - cache.num_trailing_points`)?
```suggestion
/// Apply [`flatline_check`] to a whole [`DataCache`]
///
/// Each row of `DataCache.data` can be represented as follows:
/// |---|----------|---|
/// where the first and last sections are `DataCache.num_leading_points` and
/// `DataCache.num_trailing_points` long, respectively.
/// The actual observations to be QCed (i.e. flagged) lie in the middle section.
///
/// For each row in `DataCache.data`:
/// - If `num_points` observations are identical, the last one will be flagged as [`Flag::Fail`].
/// - If any of the last `num_points` observations are missing, it will be flagged as [`Flag::DataMissing`].
/// - Otherwise they will be flagged as [`Flag::Pass`].
///
/// As (`num_points` - 1) predecessors to each observation are needed, the [`DataCache`] provided must have
/// `num_leading_points` >= `num_points` - 1.
```

intarga replied:
fixes applied in person with Manuel

Lun4m commented on src/checks/series/flatline_check.rs:
```diff
@@ -0,0 +1,107 @@
+    }
+    let data: Vec<f32> = data.iter().map(|opt| opt.unwrap()).collect();
+
+    let base = match data.first() {
+        Some(base) => base,
+        None => return Flag::Invalid,
+    };
```
I think I'd rather have an explicit `if data.is_empty()` at the start of the function and unwrap here? What do you think? Maybe I've been working a bit too much with Golang lately :rofl:

intarga replied:
After in person conversation, we decided the original is clearer

Lun4m commented on src/checks/series/flatline_check.rs:
```diff
@@ -0,0 +1,107 @@
+    num_points: u8,
+) -> Result<Vec<(String, Vec<Flag>)>, Error> {
+    let num_series = cache.data.len();
+    let mut result_vec = Vec::with_capacity(cache.data.len());
+    let series_len = match cache.data.first() {
+        Some(ts) => ts.1.len(),
+        // if this is none, the cache is empty, so we can just return an empty result vec
+        None => return Ok(result_vec),
+    };
+
```
Same as in `flatline_check`, up to your preference. 
But maybe if `data` is empty we should return an `Err` instead?

intarga replied:
We decided in person, it's fine to leave it as is

Lun4m commented on src/util/mod.rs:
```diff
@@ -36,58 +37,78 @@ pub enum Flag {
+    /// identifier, with its data points in chronological order.
+    /// All these timeseries are aligned on start_time and period.
+    /// `None`s represent gaps in the series.
+    pub data: Vec<(String, Vec<Option<f32>>)>,
```
Can this be a `Vec<Data<Option<f32>>>`? Where `Data` is something like
```rust
#[derive(Debug, Clone, PartialEq)]
pub struct Data<T> {
    pub tag: String,
    pub values: Vec<T>,
}
```
I kinda struggle with tuples, but you can make the call :)

intarga replied:
Done +extra

Lun4m commented on src/checks/series/flatline_check.rs:
```diff
@@ -0,0 +1,107 @@
+        None => return Ok(result_vec),
+    };
+
+    let (leading_trim, lead_overflow) = cache
+        .num_leading_points
+        .overflowing_sub(num_points.saturating_sub(1));
+
+    if lead_overflow || (leading_trim + num_points) as usize > series_len {
+        // TODO: nicer error here?
+        return Err(Error::InvalidInputShape("data".to_string()));
+    }
```
It's probably a waste of bytes, but maybe we should directly store `num_points`, `num_leading_points`, and `num_trailing_points` as `usize`? And also prevent `num_points == 0`? Because the check wouldn't make any sense if that were the case, no?

intarga replied:
After conversation, wontfix

Lun4m commented on src/checks/series/flatline_check.rs:
```diff
@@ -0,0 +1,107 @@
+            windows.map(flatline_check).collect(),
+        ));
+    }
+
```
If you wanted to, you could rewrite this with iterators (`cache.data.iter() ...`). Also applies to all the other checks that take a `DataCache` as argument I think

intarga replied:
After conversation, we're happy leaving this as a for loop

Lun4m commented on src/checks/series/flatline_check.rs:
```diff
@@ -0,0 +1,107 @@
+        Some(base) => base,
+        None => return Flag::Invalid,
+    };
+    if !data.iter().any(|x| x != base) {
```
Do we need to take into account `f32` shenanigans here?

intarga replied:
done

Lun4m commented on src/checks/series/monotonic_increase_check.rs:
```diff
@@ -0,0 +1,60 @@
+use crate::Flag;
+
+// TODO: why the hardcoded values?
+// TODO: one flag is returned from this, does it apply to all 25 data points?
```
Is it only for hourly data over a day maybe (current data point + previous 24 hours)?

intarga replied:
To be resolved in conversation with WP5 team

Lun4m commented on src/checks/series/monotonic_increase_check.rs:
```diff
@@ -0,0 +1,60 @@
+
+    let total_diff = data[24].unwrap() - data[0].unwrap();
+
+    if (no_decrease) && (0.7..=100.).contains(&total_diff) {
```
Yeah, this `0.7..=100.` range is magic, we need to find out where it comes from  :sweat_smile:

Lun4m commented on src/checks/series/monotonic_increase_check.rs:
```diff
@@ -0,0 +1,60 @@
+            Flag::Fail
+        );
+    }
+}
```
Can we split these into two separate tests? Or move the asserts close to their sequence defs.

intarga replied:
done

Lun4m commented on src/checks/series/spike_check.rs:
```diff
@@ -0,0 +1,137 @@
+    if data.contains(&None) {
+        return Flag::DataMissing;
+    }
+    let data: Vec<f32> = data.iter().map(|opt| opt.unwrap()).collect();
```
Would something like this be better or worse, opinions? I like it better because we don't have to deal with the indices later on
```suggestion
    let data: Vec<f32> = data.iter().map(|opt| opt.unwrap()).collect();
        let (right, obs, left) = match data[..] {
        [r, o, l] => (r.unwrap(), o.unwrap(), l.unwrap()),
        _ => unreachable!(),
    };
```

intarga replied:
done

Lun4m commented on src/checks/series/spike_check.rs:
```diff
@@ -0,0 +1,137 @@
+/// is computed. the observation will be flagged as follows
+/// - If values are missing for the observation or either neighbour: DataMissing.
+/// - If the difference is less than 35% of the sum AND the sum is greater than `max`: Fail.
+/// - If the difference is less than 35% of the sum AND the sum is greater than `high`: Warn.
```
But there is no `Flag::Warn` in the check, maybe it's TODO?

intarga replied:
fixed

Lun4m commented on src/checks/series/spike_check.rs:
```diff
@@ -0,0 +1,137 @@
+
+/// Apply [`spike_check`] to a whole [`DataCache`]
+///
+/// As a predecessor and successor to each observation are needed, the [`SeriesCache`] provided
```
```suggestion
/// As a predecessor and successor to each observation are needed, the [`DataCache`] provided
```

intarga replied:
fixed

Lun4m commented on src/checks/series/step_check.rs:
```diff
@@ -0,0 +1,114 @@
+
+    if (data[0] - data[1]).abs() > max {
+        return Flag::Fail;
+    }
```
Probably it doesn't make any difference?
```suggestion
    if (data[0].unwrap() - data[1].unwrap()).abs() > max {
        return Flag::Fail;
    }
```

intarga replied:
done

Lun4m commented on src/checks/series/step_check.rs:
```diff
@@ -0,0 +1,114 @@
+/// If the absolute value of the difference between the observed value and it's predecessor is
+/// greater than max, Flag::Fail will be returned for that observation, if greater than high,
+/// Flag::Warn, if either value if missing, Flag::DataMissing, else Flag::Pass.
+///
```
Maybe we can be a bit more systematic with the docs? Something like
```suggestion
/// Timeseries QC test that compares each observation against its immediate predecessor.
/// 
/// Returns:
/// - [`Flag::DataMissing`], if either value is missing,
/// - [`Flag::Warn`], if the difference is greater than `high`,
/// - [`Flag::Fail`], if the difference is greater than `max`,
/// - [`Flag::Pass`], otherwise.
///
```
 and keep the same style for every check (as much as possible, of course).
Also, did we decide not to return `Flag::Warn`?

intarga replied:
done

Lun4m commented on src/checks/series/step_check.rs:
```diff
@@ -0,0 +1,114 @@
+///
+/// As a predecessor and successor to each observation are needed, the [`SeriesCache`] provided
+/// must have `num_leading_points` and `num_trailing_points` >= 1. A constant is provided to aid
+/// in enforcing this constraint
```
`SeriesCache` &rarr; `DataCache`

intarga replied:
done

Lun4m commented on src/checks/single/range_check.rs:
```diff
@@ -0,0 +1,85 @@
+    cache: &DataCache,
+    upper_limit: f32,
+    lower_limit: f32,
+) -> Vec<(String, Vec<Flag>)> {
```
Should these single checks that take a `DataCache` also return a `Result`? I've already forgotten how we use these in Rove :sweat_smile:

intarga replied:
After conversation, wontfix

Lun4m commented on src/checks/single/range_check_humidity.rs:
```diff
@@ -0,0 +1,83 @@
+//TODO: is this the optimal return signature for corrections?
+/// Apply [`range_check_humidity`] to a whole [`DataCache`]
+#[allow(clippy::type_complexity)]
+pub fn range_check_humidity_cache(cache: &DataCache) -> Vec<(String, Vec<(Flag, Option<f32>)>)> {
```
We could use the same struct I mentioned in `util/mod.rs`

intarga replied:
done

Lun4m commented on src/checks/single/range_check_wind_direction.rs:
```diff
@@ -0,0 +1,97 @@
+
+/// Range check with a correction for wind direction outside 0-360.
+///
+/// If the direction is -20-0, or 360-380, 360 will be added or subtracted to get it back into
```
```suggestion
/// If the direction is in [-20, 0), or [360, 380], 360 will be added or subtracted to get it back into
```
Range notation?

intarga replied:
obsoleted

Lun4m commented on src/checks/single/range_check_wind_direction.rs:
```diff
@@ -0,0 +1,97 @@
+/// If the direction is -20-0, or 360-380, 360 will be added or subtracted to get it back into
+/// the correct range.
+pub fn range_check_wind_direction(datum: Option<f32>) -> (Flag, Option<f32>) {
+    // TODO: get to the bottom of weird -3.0 handling: kvalobs code looks for a value -3.0, and
+    // avoids flagging that if X_5 (lowest?) is also -3.0. From comments in the code, it looks like
+    // this has to do with a special param_id?
```
Does it mean that this check needs also to take `min` and `max` arguments? But yeah, sounds a bit weird it's only checking for exactly `-3.0`

intarga replied:
Clarified in a conversation

Lun4m commented on src/checks/spatial/buddy_check.rs:
```diff
@@ -1,4 +1,29 @@
+/// signature more readable.
+#[derive(Debug, Clone)]
+pub struct BuddyCheckArgs {
+    /// Search radius in which buddies of an observation will be found. Unit: m
+    pub radii: SingleOrVec<f32>,
+    /// The minimum buddies an observation can have (lest it be flagged [`Flag::Isolated`])
+    pub nums_min: SingleOrVec<u32>,
```
```suggestion
    /// Search radius in which buddies of an observation will be found. Unit: m
    pub radii: SingleOrVec<f32>,
    /// The minimum buddies an observation can have (lest it be flagged [`Flag::Isolated`])
    pub min_buddies: SingleOrVec<u32>,
```
Buddies sounds cool :smile: But why would these be `SingleOrVec` and not just the inner type (assuming we are QC'ing obses from the same timeseries)? Or are we also QC'ing the backing data?

intarga replied:
Clarified in conversation

Lun4m commented on src/checks/spatial/buddy_check.rs:
```diff
@@ -152,6 +154,34 @@ pub fn buddy_check(
     Ok(flags)
 }
 
+/// Apply [`buddy_check`] to a whole [`DataCache`]
```
I need a visual diagram :sweat_smile:

intarga replied:
clarified in conversation

Lun4m commented on src/checks/spatial/buddy_check.rs:
```diff
@@ -152,6 +154,34 @@ pub fn buddy_check(
+) -> Result<Vec<(String, Vec<Flag>)>, Error> {
+    let series_len = cache.data[0].1.len();
+
+    let mut result_vec: Vec<(String, Vec<Flag>)> = cache
+        .data
+        .iter()
+        .map(|ts| (ts.0.clone(), Vec::with_capacity(series_len)))
+        .collect();
+
+    for i in (cache.num_leading_points as usize)..(series_len - cache.num_trailing_points as usize)
+    {
+        let timeslice: Vec<Option<f32>> = cache.data.iter().map(|v| v.1[i]).collect();
+        let spatial_result = buddy_check(&timeslice, &cache.rtree, args, obs_to_check)?;
+
+        for i in 0..spatial_result.len() {
+            result_vec[i].1.push(spatial_result[i]);
+        }
+    }
+
+    Ok(result_vec)
```
```suggestion
    let result_vec = ((cache.num_leading_points as usize)
        ..(series_len - cache.num_trailing_points as usize))
        .map(|i| {
            let timeslice: Vec<Option<f32>> = cache.data.iter().map(|v| v.1[i]).collect();
            Ok((
                cache.data[0].0.clone(),
                buddy_check(&timeslice, &cache.rtree, args, obs_to_check)?,
            ))
        })
        .collect();

    result_vec
```
But I'm not sure if the `String`s are right the way they are (in either version)

intarga replied:
after conversation, wontfix

Lun4m commented on src/checks/spatial/buddy_check.rs:
```diff
@@ -161,7 +191,20 @@ mod tests {
         assert_eq!(
             buddy_check(
-                &SpatialCache::new(
+                &[
+                    Some(0.),
+                    Some(0.),
+                    Some(0.),
+                    Some(0.),
+                    Some(0.),
+                    Some(0.),
+                    Some(0.),
+                    Some(0.),
+                    Some(0.1),
+                    Some(1.)
+                ]
+                .to_vec(),
```
Clippy suggests using `as_ref()` instead of `to_vec()`

intarga replied:
fixed

Lun4m reviewed (changes requested):
I still haven't fully looked into `buddy_check` and `src` :smiling_face_with_tear:

intarga committed:
fix min/max typos in doc comments

intarga committed:
remove reference to `QC test` in flatline_check docs

intarga committed:
enable clippy linting of tests in CI

(and fix lints we missed)

intarga committed:
replace collecting unwrapped vals to vector with just unwrap

in step_check, spike_check, and flatline_check

intarga committed:
clarify docs on DataCache

intarga committed:
make format for docs of series checks more consistent

intarga committed:
make format of docs for single checks more consistent

intarga committed:
make format of docs for consistency checks more consistent

intarga committed:
introduce Timeseries abstraction to simplify types

intarga committed:
add a threshold parameter for acceptable difference in flatline_check

intarga committed:
reorder asserts in unit test for monotonic_increase_check for clarity

intarga committed:
rename parameter nums_min to min_buddies in buddy_check for aesthetics
